### PR TITLE
Fix crash with 'use conditional' in a VB property

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/29376

--

VB IOperation tree is 'interesting' for properties.  There is an implicit local declaration as the first statement in the body, containing a variable.  Reads/writes to that property are actually represented in the iop tree as reads/writes to that variable.  This threw off 'use conditional' which then wanted to merge the conditional into that local declaration statement.  Of course, since this wasn't a real local declaration statement, this blew up.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
tagging @dotnet/roslyn-ide @jcouv @mavasani 
This fixes a customer reported crash.

File: src/Features/Core/Portable/UseConditionalExpression/ForAssignment/AbstractUseConditionalExpressionForAssignmentCodeFixProvider.cs [hunk 1]
```diff
@@ -168,6 +168,11 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
                 return false;
             }
 
+            if (localDeclaration.IsImplicit)
+            {
+                return false;
+            }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UseConditionalExpression/ForAssignment/AbstractUseConditionalExpressionForAssignmentCodeFixProvider.cs]: this code is in the TryFindMatchingLocalDeclarationImmediatelyAbove method.  THe purpose of the code is to find a *real* user local-decl statement right above hte if-statement so we can merge our result into it.  It was previously returning implicit statements (which VB creates), leading to bad baddy badness.

**Comment by CyrusNajmabadi** [comment]
@jinujoseph Can i get a buddy for thsi crash fix?

**Comment by CyrusNajmabadi** [comment]
@mavasani can you take a look?  Thanks!

**Comment by AlekseyTs** [comment]
> Reads/writes to that property are actually represented in the iop tree as reads/writes to that variable.

These are not read/writes to the property. These are read/writes to the implicitly declared function local, which has the same name as the property. IOperation tree reflects that. I don't know why a tool-tip in IDE says the target of the assignments is a property, feels like a bug to me.

File: src/EditorFeatures/VisualBasicTest/UseConditionalExpression/UseConditionalExpressionForAssignmentTests.vb [hunk 2]
```diff
@@ -547,6 +547,36 @@ class C
...
+        Public Async Function TestOnAssignmentToContainingProperty() As Task
+            Await TestInRegularAndScriptAsync(
+"
+class C
+    public theemployee as string
+
+    public readonly property employee as string
+        get
+            [||]if theemployee is nothing then
+                employee = ""1""
```
Comment (AlekseyTs) [inline src/EditorFeatures/VisualBasicTest/UseConditionalExpression/UseConditionalExpressionForAssignmentTests.vb]: >employee = ""1"" [](start = 16, length = 16)

Consider adding similar test for function local inside a regular function. 
```
Class C
    Public theemployee As String

    Public Function employee() As String
        If theemployee Is Nothing Then
            employee = "1"
        Else
            employee = "2"
        End If
    End Function
```

Comment (AlekseyTs) [inline src/EditorFeatures/VisualBasicTest/UseConditionalExpression/UseConditionalExpressionForAssignmentTests.vb]: >employee = ""1"" [](start = 16, length = 16)

Consider adding similar test for an implicitly declared local: 

```
Option Explicit Off

Class C
    Public theemployee As String

    Public Sub F()
        If theemployee Is Nothing Then
            employee = "1"
        Else
            employee = "2"
        End If
    End Sub
End Class
```

**Comment by CyrusNajmabadi** [comment]
Will add the test suggestions. Thanks for them @AlekseyTs !

**Comment by CyrusNajmabadi** [comment]
> I don't know why a tool-tip in IDE says the target of the assignments is a property, feels like a bug to me. SemanticModel says it is a local.

IIRC, this has been very old behavior in roslyn from very early on.  Either the semantic model returns things as if it's hte property (and the IDE just shows it that way), or hte semantic model returns the right thing, however for some reason the IDE chooses to override it (maybe legacy?).  

I know this issue is one of those interesting historical VBisms htat tends to bite a lot of features since no one tends to think about it much.

**Comment by CyrusNajmabadi** [comment]
@mavasani can you wait on merging this in so i can add the tests Aleksey mentioned?  I'll let you know when they're in.  Thanks!

**Comment by CyrusNajmabadi** [comment]
@mavasani tests were added.  this can go in whenever you want.

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**